Implement restore of DSCP on egress

dave14305 opened the pull request:
Also consolidate the forward and output chains into postrouting.

Improve the hotplug script to only restart if the service is enabled.

Cleanup trailing whitespace throughout.

dave14305 committed:
Only invoke hotplug if service is enabled

dave14305 committed:
Cleanup stray whitespace

dave14305 committed:
Restore DSCP on egress to avoid nftables chains

After the initial packet in a conversation has been marked and saved
to conntrack mark, avoid processing additional packets from the
same conversation through nftables and directly restore the DSCP
on egress to the WAN interface.

Before entering the classification chains, ensure the ct mark 128 bit
mask has not been set yet.

dave14305 committed:
Move hook to postrouting

The postrouting hook will see both output and forward traffic, so
combine the chains into one for simplicity, since the same rules
and maps apply to both chains currently.

lynxthecat commented on cake-qos-simple:
```diff
@@ -11,9 +11,9 @@
 
 # This facilitates the optional use of the diffserv functionality in cake for improving qos when
-# bandwidth is contrained is contrained by leveraging the connection trability in Linux. 
+# bandwidth is contrained is contrained by leveraging the connection trability in Linux.
```
```suggestion
# bandwidth is constrained by leveraging connection tracking (conntrack) in Linux.
```

lynxthecat commented:
@rany2 do you know whether I can amend those commit messages associated with this pull request myself or whether @dave14305 has to?

lynxthecat committed:
Resolve typos in comment

lynxthecat commented on cake-qos-simple:
```diff
@@ -76,6 +76,9 @@ start()
 	tc qdisc add dev "${ul_if}" root handle 1: cake bandwidth "${cake_ul_rate_Mbps}Mbit" ${cake_ul_options}
 
+	printf "\nSetting up tc filter to restore DSCPs from conntrack on egress packets on interface '${ul_if}'.\n"
+	tc filter add dev "${ul_if}" matchall action ctinfo dscp 63 128
```
Should we make this a higher priority than the ECN filter below so that the ECN filter will apply second?

dave14305 replied:
I’m not sure it matters since they would be mutually exclusive of each other. Restoring DSCP doesn’t change ECN bits, or vice versa.

I haven’t used that feature since I haven’t built pedit into my custom build yet.

lynxthecat replied:
Any clue what would be the syntax to make this 'prio 1' and the other 'prio 2'? The pedit lines are the ones I can't seem to figure out.

lynxthecat commented on cake-qos-simple:
```diff
@@ -76,6 +76,9 @@ start()
 	tc qdisc add dev "${ul_if}" root handle 1: cake bandwidth "${cake_ul_rate_Mbps}Mbit" ${cake_ul_options}
 
+	printf "\nSetting up tc filter to restore DSCPs from conntrack on egress packets on interface '${ul_if}'.\n"
+	tc filter add dev "${ul_if}" matchall action ctinfo dscp 63 128
```
```suggestion
	tc filter add dev "${ul_if}" prio 1 matchall action ctinfo dscp 63 128
```

lynxthecat committed:
Explicitly specify parent and protocol

lynxthecat commented:
@dave14305 do you agree with https://github.com/lynxthecat/cake-qos-simple/pull/6/commits/5dd71935fa1373cd2b790f78f72b52a546fff733?

dave14305 commented:
> @dave14305 do you agree with [5dd7193](https://github.com/lynxthecat/cake-qos-simple/commit/5dd71935fa1373cd2b790f78f72b52a546fff733)?

I’m not bothered by it.

lynxthecat committed:
Prioritise DSCP restore over ECN overwrite

dave14305 commented:
One more tweak might be to move the chain from `priority filter` to `priority mangle + 1` to ensure the save-and-store occurs after any fw4 DSCP rules.

dave14305 committed:
Move hook-postrouting chain to mangle priority

Any user-configured fw4 rules that change DSCP will be written to
fw4's mangle_postrouting chain. For consistemcy, move this chain to
mangle priority + 1 so that it runs after the fw4 rules.

Technically filter priority would have been fine, but generally any
packet mangling should happen in the mangle portion of the rule traversal.
